firmware/Sources/HAL/HAL_ADC/hal_adc.c

```diff
@@ -0,0 +1,166 @@
+static volatile uint8_t _idx_cplt_ADC2 = _ADC2_BUFFER_SIZE/2;
+
+static	uint16_t _meas_pwr[_N_SAMPLES_PWR];
+
+/**********************************************************************************/
+/*                        Definition of exported variables                        */
+/**********************************************************************************/
+extern ADC_HandleTypeDef hadc1;
+extern ADC_HandleTypeDef hadc2;
+extern uint8_t EPC_ST_ERR_COUNTER;
+
+/**********************************************************************************/
+/*                      Definition of exported constant data                      */
+/**********************************************************************************/
+
+/**********************************************************************************/
+/*                    Declaration of local function prototypes                    */
+/**********************************************************************************/
+
+/**********************************************************************************/
+/*                       Definition of local constant data                        */
+/**********************************************************************************/
+
+/**********************************************************************************/
+/*                         Definition of local functions                          */
+/**********************************************************************************/
+
+// Inserts a key in arr[] of given capacity. n is current
+// size of arr[]. This function returns n+1 if insertion
+// is successful, else n.
+static uint16_t insertSorted(const uint16_t n, uint16_t key)
+{
+    // Cannot insert more elements if n is already
+    // more than or equal to capacity
+		uint16_t new_size = n;
+		if (n < _N_SAMPLES_PWR) {
+			new_size += 1;
+			if (n == 0){
+				_meas_pwr[0] = key;
+			} else{
+				int16_t i = n - 1;
+				for (; (i >= 0 && _meas_pwr[i] > key); i--)
+					_meas_pwr[i + 1] = _meas_pwr[i];
+
+				_meas_pwr[i + 1] = key;
+			}
+    }
+
+    return new_size;
+}
+
+
+// Function for calculating median
+static uint16_t getPwrMedian(uint16_t start_sensor_idx)
+{
+	// Insert data ordered
+	uint16_t meas_idx = start_sensor_idx;
+	uint8_t stride = 3, i;
+	for(i = 0; i < _N_SAMPLES_PWR;){
+		i = insertSorted(i, _ADC2_results[meas_idx]);
+		meas_idx += stride;
+	}
+
+	// check for even case
+  uint16_t res = 0;
+	if (i % 2 != 0)
+		res = _meas_pwr[i / 2];
+	else
+		res = (_meas_pwr[(i - 1) / 2] + _meas_pwr[i / 2]) / 2;
+  return res;
+}
+
+
+/**********************************************************************************/
+/*                        Definition of exported functions                        */
+/**********************************************************************************/
+void HAL_Adc1_Callback(void){
+	_idx_cplt_ADC1 = _ADC1_BUFFER_SIZE/2 - _idx_cplt_ADC1;
+}
+
+void HAL_Adc2_Callback(void){
```

Comment:
is possible to follow naming convention?

Reply:
Done